File: core/src/net/brenig/pixelescape/game/data/UserData.java
package net.brenig.pixelescape.game.data;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Preferences;

import net.brenig.pixelescape.lib.LogHelper;
import net.brenig.pixelescape.lib.Names;
import net.brenig.pixelescape.lib.Reference;

/**
 * Used to store user data, such as highscore
 */
public class UserData {

	private static final class Keys {
		public static final String savesRevision = "data_version";

		public static final String highScore = "highscore";
		public static final String gameMode = "gamemode";

		public static final String tutorial = "tutorial";
	}

	private static final class Defaults {
		private static final int highScore = 0;
		private static final int gameMode = 0;
		private static final boolean tutorial = false;
	}

	private static final String PREF_MAIN_DATA = "PixelEscape_User_Data";


	private final Preferences prefs;

	public UserData() {
		prefs = Gdx.app.getPreferences(PREF_MAIN_DATA);
	}

	public void saveToDisk() {
		prefs.flush();
		LogHelper.log("UserData", "User progress saved!");
	}

	/**
	 * updates the HighScore
	 * @param gameMode the current GameMode
	 * @param score the new score
	 * @return true if the score got updated (--> new highscore)
	 */
	public boolean updateHighscore(GameMode gameMode, int score) {
		int lastHighScore = getHighScore(gameMode);
		if(score > lastHighScore) {
			setHighScore(gameMode, score);
			return true;
		}
		return false;
	}

	/**
	 * @param gameMode the current GameMode
	 * @return current HighScore for the given GameMode
	 */
	public int getHighScore(GameMode gameMode) {
		return prefs.getInteger(Keys.highScore + gameMode.getScoreboardName(), Defaults.highScore);
	}

	/**
	 * Sets the new HighScore for the given GameMode
	 */
	public void setHighScore(GameMode gameMode, int highScore) {
		prefs.putInteger(Keys.highScore + gameMode.getScoreboardName(), highScore);
		saveToDisk();
	}


	/**
	 * @return the latest selected gamemode
	 */
	public int getLastGameMode() {
		return prefs.getInteger(Keys.gameMode, Defaults.gameMode);
	}

	public void setLastGameMode(int id) {
		prefs.putInteger(Keys.gameMode, id);
		saveToDisk();
	}

	public boolean tutorialSeen(GameMode gameMode) {
		if(GameDebugSettings.get("FORCE_TUTORIALS")) {
			return true;
		}
		//noinspection PointlessBooleanExpression,ConstantConditions
		return Reference.SUSPRESS_TUTORIALS || prefs.getBoolean(Keys.tutorial + gameMode.getGameModeName(), Defaults.tutorial);
	}

	public void setTutorialSeen(GameMode gameMode, boolean value) {
		prefs.putBoolean(Keys.tutorial + gameMode.getGameModeName(), value);
	}

	/**
	 * updates exsisting data to new format if necessary
	 */
	public void updateSaveGames() {
		int savedDataVersion = prefs.getInteger(Keys.savesRevision, -1);
		if(savedDataVersion < Reference.PREFS_REVISION) {
			LogHelper.log("UserData", "converting old save data!");
			switch (savedDataVersion) {
				case -1:
					int oldHighScore = prefs.getInteger(Keys.highScore, Defaults.highScore);
					prefs.putInteger(Keys.highScore + Names.SCOREBOARD_CLASSIC, oldHighScore);
				case 1:
					prefs.putInteger(Keys.highScore + Names.SCOREBOARD_ARCADE, 0);
			}
			prefs.putInteger(Keys.savesRevision, Reference.PREFS_REVISION);
			saveToDisk();
		}
	}

}

File: core/src/net/brenig/pixelescape/game/data/UserData.kt
package net.brenig.pixelescape.game.data

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Preferences
import net.brenig.pixelescape.game.data.constants.Reference
import net.brenig.pixelescape.game.data.constants.ScoreboardNames
import net.brenig.pixelescape.lib.log

/**
 * Used to store user data, such as highscore
 */
class UserData {

    private val prefs: Preferences


    /**
     * @return the latest selected gamemode
     */
    var lastGameMode: Int
        get() = prefs.getInteger(Keys.gameMode, Defaults.gameMode)
        set(id) {
            prefs.putInteger(Keys.gameMode, id)
            saveToDisk()
        }

    init {
        prefs = Gdx.app.getPreferences(PREF_MAIN_DATA)
    }

    fun saveToDisk() {
        prefs.flush()
        log("UserData", "User progress saved!")
    }

    /**
     * updates the HighScore
     *
     * @param gameMode the current GameMode
     * @param score    the new score
     * @return true if the score got updated (--> new highscore)
     */
    fun updateHighscore(gameMode: GameMode, score: Int): Boolean {
        val lastHighScore = getHighScore(gameMode)
        if (score > lastHighScore) {
            setHighScore(gameMode, score)
            return true
        }
        return false
    }

    /**
     * @param gameMode the current GameMode
     * @return current HighScore for the given GameMode
     */
    fun getHighScore(gameMode: GameMode): Int {
        return prefs.getInteger(Keys.highScore + gameMode.scoreboardName, Defaults.highScore)
    }

    /**
     * Sets the new HighScore for the given GameMode
     */
    fun setHighScore(gameMode: GameMode, highScore: Int) {
        prefs.putInteger(Keys.highScore + gameMode.scoreboardName, highScore)
        saveToDisk()
    }

    fun tutorialSeen(gameMode: GameMode): Boolean {
        return if (GameDebugSettings["FORCE_TUTORIALS"]) {
            true
        } else {
            !Reference.ENABLE_TUTORIALS || prefs.getBoolean(Keys.tutorial + gameMode.gameModeName, Defaults.tutorial)
        }
    }

    fun setTutorialSeen(gameMode: GameMode, value: Boolean) {
        prefs.putBoolean(Keys.tutorial + gameMode.gameModeName, value)
    }

    /**
     * updates existing data to new format if necessary
     */
    fun updateSaveGames() {
        val savedDataVersion = prefs.getInteger(Keys.savesRevision, -1)
        if (savedDataVersion < Reference.PREFS_REVISION) {
            log("UserData", "converting old save data!")
            when (savedDataVersion) {
                -1 -> {
                    val oldHighScore = prefs.getInteger(Keys.highScore, Defaults.highScore)
                    prefs.putInteger(Keys.highScore + ScoreboardNames.SCOREBOARD_CLASSIC, oldHighScore)
                    prefs.putInteger(Keys.highScore + ScoreboardNames.SCOREBOARD_ARCADE, 0)
                }
                1 -> prefs.putInteger(Keys.highScore + ScoreboardNames.SCOREBOARD_ARCADE, 0)
            }
            prefs.putInteger(Keys.savesRevision, Reference.PREFS_REVISION)
            saveToDisk()
        }
    }

    private object Keys {
        const val savesRevision = "data_version"

        const val highScore = "highscore"
        const val gameMode = "gamemode"

        const val tutorial = "tutorial"
    }

    private object Defaults {
        const val highScore = 0
        const val gameMode = 0
        const val tutorial = false
    }

    companion object {
        private const val PREF_MAIN_DATA = "PixelEscape_User_Data"
    }

}
